tests/e2e/datadog/metrics/series.rs

```diff
@@ -0,0 +1,278 @@
+    fn from(input: &DatadogSeriesMetric) -> Self {
+        let mut resources = vec![];
+        if let Some(host) = &input.host {
+            resources.push(Resource {
+                r#type: "host".to_string(),
+                name: host.clone(),
+            });
+        }
+
+        let points = input
+            .points
+            .iter()
+            .map(|point| MetricPoint {
+                value: point.1,
+                timestamp: point.0,
+            })
+            .collect();
+
+        let interval = input.interval.unwrap_or(0) as i64;
+
+        let r#type = match input.r#type {
+            vector::common::datadog::DatadogMetricType::Gauge => 3,
+            vector::common::datadog::DatadogMetricType::Count => 1,
+            vector::common::datadog::DatadogMetricType::Rate => 2,
+        };
+
+        MetricSeries {
+            resources,
+            metric: input.metric.clone(),
+            tags: input.tags.clone().unwrap_or_default(),
+            points,
+            r#type,
+            unit: "".to_string(),
+            source_type_name: input.clone().source_type_name.unwrap_or_default(),
+            interval,
+            metadata: None,
+        }
+    }
+}
+
+fn convert_v1_payloads_v2(input: &[DatadogSeriesMetric]) -> Vec<MetricPayload> {
+    input
+        .iter()
+        .map(|serie| MetricPayload {
+            series: vec![serie.into()],
+        })
+        .collect()
+}
+
+fn unpack_v1_series(in_payloads: &[FakeIntakePayloadJson]) -> Vec<DatadogSeriesMetric> {
+    in_payloads
+        .iter()
+        .flat_map(|payload| {
+            let series = payload.data.as_array().unwrap();
+            series
+                .iter()
+                .map(|serie| serde_json::from_value(serie.clone()).unwrap())
+        })
+        .collect()
+}
+
+async fn get_v1_series_from_pipeline(address: String) -> SeriesIntake {
+    info!("getting v1 series payloads");
+    let payloads =
+        get_fakeintake_payloads::<FakeIntakeResponseJson>(&address, SERIES_ENDPOINT_V1).await;
+
+    info!("unpacking payloads");
+    let payloads = unpack_v1_series(&payloads.payloads);
+    info!("converting payloads");
+    let payloads = convert_v1_payloads_v2(&payloads);
+
+    info!("generating series intake");
+    let intake = generate_series_intake(&payloads);
+
+    common_series_assertions(&intake);
+
+    info!("{:?}", intake);
+
+    intake
+}
+
+async fn get_v2_series_from_pipeline(address: String) -> SeriesIntake {
+    info!("getting v2 series payloads");
+    let payloads =
+        get_fakeintake_payloads::<FakeIntakeResponseRaw>(&address, SERIES_ENDPOINT_V2).await;
+
+    info!("unpacking payloads");
+    let payloads = unpack_proto_payloads::<MetricPayload>(&payloads);
+
+    info!("generating series intake");
+    let intake = generate_series_intake(&payloads);
+
+    common_series_assertions(&intake);
+
+    info!("{:?}", intake);
+
+    intake
+}
+
+pub(super) async fn validate() {
```

Comment:
I want to check my understanding. The assertions we are making here are (1) the last gauge values are equal (e.g., `50` given the current test script) and (2) the sum of the rates over the course of the entire test are equal. Is that correct?

Reply:
Mostly correct.

I tried to describe this a little bit on line 50 with the doc comment on `SeriesIntake`. Let me know if it would be better to add some comments to this function.

There are specific assertions for the different metric types, but also a higher level assertion in the consistency of the shape of the data over time, regardless of metric type.

For the specific assertions for different metric types, your (1) and (2) assertions are correct with the slight adjustment that those assertions apply on each request that was received (potentially containing more than one points value).

The higher level assertion on the consistency of the shape of the data over time is provided through the `TimeSeriesData` structure. The organization of the data using a BTreeMap, and the comparison that starts on line 236, allows us to ensure that the general shape of data that is output from Vector over time, is the same as that of the Agent case. That is, our data points in the time series line up with the Agents, using the TimeBucket to abstract the actual timestamp values, but maintain the relative order in the time series.

Reply:
> For the specific assertions for different metric types, your (1) and (2) assertions are correct with the slight adjustment that those assertions apply on each request that was received (potentially containing more than one points value).

Thanks, I see. Given that these comparisons are made per request, will this result in flakey tests if the metrics are batched into requests differently in the agent and vector?

Reply:
Gah, sorry, I misused the "on each request".

What's actually happening is that we are essentially collapsing the batching behavior as part of the processing into the `SeriesIntake` data structure. We map based on timestamp. So any data points that come in on any batch will still land in the correct timestamp bucket irrespective of batch order or if there is a different number of batches in Agent vs Vector.

When I said "with the slight adjustment that those assertions apply on each request that was received " , what I meant was "with the slight adjustment that those assertions apply on each timestamp that was received ".

In my local runs I've yet to encounter flakiness.

Reply:
I attempted to write this up in https://github.com/vectordotdev/vector/pull/18840/commits/031cdcc6d99d96e8113df54c9f5fc1e286c02d39

Reply:
Nice, the code comment explains this all nicely. Thank you for adding that!